sign/ed25519/ed25519.go

```diff
@@ -197,9 +188,13 @@ func sign(kp *KeyPair, message []byte, preHash bool, ctx []byte) ([]byte, error)
 // The opts.HashFunc() must return SHA512 to indicate the message has been
 // hashed with SHA512. This can be achieved by passing crypto.SHA512 as the value
 // for opts.
+// Context should be passed to this function, which length should be more than
+// 255.
```

Comment:
```suggestion
// Context should be passed to this function, which length should be no more than
// 255.
```

I think this applies elsewhere where the length of `ctx` is encoded in a single byte, right?

Reply:
Oh, yes! That is right. I'll change ;)

Reply:
Just fixed it. I also fixed some context checking, as `SignPhCtx` should be able to accept an empty context. The RFC only states: "For Ed25519ph, phflag=1 and PH is SHA512 instead.  That is, the input is hashed using SHA-512 before signing with Ed25519"; but not restriction on the context.. so I assume that an empty string can be passed ;) But let me know if you think otherwise, as it is not the clearest thing on the RFC ;)

Thanks!

Reply:
There are only three variants:

| Name in RFC | Prehash | Valid Context `len(ctx)` | Name in Code |
|-----|-----|-----|-----|
| Ed25519 | No | No - None | SignPure or Sign |
| Ed25519ctx | No | Yes - (1-255) | SignWithCtx |
| Ed25519ph | Yes | Yes - (0-255) | SignPh | 

That leads to the following declarations:
` Sign(message []byte, opts crypto.SignerOpts)  []byte`
` SignWithCtx(message []byte, ctx []byte)  []byte`
` SignPh(message []byte, ctx []byte)  []byte`
and
` Verify(publicKey PublicKey, message, sig []byte) bool`
` VerifyWithCtx(publicKey PublicKey, message, ctx, sig []byte) bool`
` VerifyPh(publicKey PublicKey, message, ctx, sig []byte) bool`

Now, the special case to support the `crypto.Signer` interface can only call to `SignPure` because calling to `SignPh` still requires of passing a context, which is not included in the `crypto.SignerOpts`.

Reply:
Thanks for this @armfazh ! So, the reason why the `crypto.Signer` interface one calls now both to `SignPure` and `SignPh` is due to the golang proposal where this was what the accepted api will look like. I can also used the method described here: https://github.com/golang/go/issues/31804#issuecomment-537230419 , and then the `crypto.Signer`  can accept context.